---
title: "ranscriptomics and the analysis of RNA-Seq data"
author: "Analine Aguayo"
date: "11/13/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
##Setup Bioconductor DESeq2

```{r eval=FALSE}
 install.packages("BiocManager")
 BiocManager::install()
 BiocManager::install("DESeq2")
 library(DESeq2)
```
```{r}
library(DESeq2)
```


```{r}
read.csv("airway_scaledcounts.csv")
read.csv("airway_metadata.csv")
```

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = F)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = F)
```

lets take a peak
```{r}
head(counts)
```
how many genes do we have in the database?

```{r}
nrow(counts)
```

how many experiments?
```{r}
nrow(counts)
```
```{r}
ncol(counts)-1

```

lets make sure metadata id col matches the colnames of counts
```{r}
colnames(counts)[-1]
```
```{r}
metadata$id
```
ˆare they the same?
```{r}
all(colnames(counts)[-1]==metadata$id)
```

```{r}
all(c(T,T,T))
```
##Analysis: compare the control to drug treated
First we need to access the columns of our countData that are control and then find their mean.
```{r}
metadata$dex
```

```{r}
metadata$dex=="control"
```

```{r}
metadata[metadata$dex=="control",]$id
#these are the blue columns that are the controls

```

```{r}
control.id <- metadata[metadata$dex=="control",]$id
control.id
```
focus on the countrs for control experiments...
```{r}
head(counts[, control.id])
```
Mean counts for the control experiments.
```{r}

control.mean <- rowSums(counts[, control.id])/length(control.id)
names(control.mean) <- counts$ensgene
```

```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/4 
names(treated.mean) <- counts$ensgene
```
Lets story these together
```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

Plot control vs mean.
```{r}
plot(meancounts$control,meancounts$treated)
```
Good case for a log plot
```{r}
plot(meancounts$control,meancounts$treated,log="xy")

```
```{r}
#in this plot the dots that are on either side of the xy line, mean that there was a differential expression of this gene depending on the group (treatment vs control)
```
Going to normalize this data by the mean.control
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```
There are a couple of “weird” results. Namely, the NaN (“not a number”“) and -Inf (negative infinity) results.

The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes. Again inspect your result (and the intermediate steps) to see if things make sense to you
```{r}
#test for finding zwro entries
x <- c(5,8,0,5)
x ==0
```

```{r}
which(x==0)
```
```{r}
x <- data.frame(happy=c(5,6,0,0), sad = c(0,5,5,0))
x ==0
```
```{r}
which(x==0, arr.ind=TRUE)
```
focus on the rows only that have zero entries
```{r}
which(x==0, arr.ind=TRUE)[,1]

```

```{r}
unique(which(x==0, arr.ind=TRUE)[,1]
)
```

```{r}
inds <-unique(which(x==0, arr.ind=TRUE)[,1])
x[-inds,]

```
now we are gonna make our neighbor try to understand my new code with different happy and sad numbers
```{r}
x <- data.frame(happy=c(5,6,4,0), sad = c(0,0,5,04))
x ==0
```
```{r}
which(x==0, arr.ind=TRUE)
```

```{r}
unique(which(x==0, arr.ind=TRUE)[,1]
)
```

```{r}
x[-inds,]
```

now do this for our menacounts data, ie remove xero containing fenes from further  consideration.
```{r}
to.rm<-unique(which(meancounts[1:2]==0, arr.ind=TRUE)[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
how many genes do i have left?
```{r}
nrow(mycounts)
```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.


```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

```{r}
sum(up.ind)
```

##Annotation

we will use Bioconductor's annotation packages to help with mapping various id schemes to each other. here we load the annotationdbi package and the annotation package org.Hs.eg.db.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```

```{r}
head(mycounts)
```

```{r}

```

